docs/design/proposals/codegen-refactor/centralize_1.md

```diff
@@ -0,0 +1,197 @@
+
+
+### Requirements
+* No impact to existing functionality
+* Code is easy to read, intuitive, and extendable
+* Naming in code is consistent (ex: Find vs. Get)
+* Navigating code is easier because pkg responsibility is clear
+
+
+### Prerequisites
+*code-generator* contains unit tests and also runs a subset of $SERVICE-controller e2e tests against PRs. Are these comprehensive enough to detect breakage in the code that we'll be touching/refactoring? If there are gaps between tests and code being refactored, then **tests will need to be created or updated** prior to implementing the solution.
+
+Agree on terms for consistent code. Terms such as Get vs. Find, CRDS/Resource, etc.
+
+### Implementation
+
+#### Consolidate `ackgenconfig` access and expose Getters
+The purpose is to realign responsibilities of `CRD` and `Config` making code more consistent and intuitive.
+
+1. Remove `ackgenconfig` as an attribute in `type CRD struct` in [crd.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/model/crd.go#L65) and fix breakage
+  * move the associated methods to `config` package and update callers, ex: [GetOutputWrapperFieldPath](https://github.com/aws-controllers-k8s/code-generator/blob/59c6892e4b61b5e2076e5e5504daba8278b82980/pkg/model/crd.go#L420):
+
+```
+// Current
+wrapperFieldPath := r.GetOutputWrapperFieldPath(op)
+
+// Proposed
+wrapperFieldPath := cfg.GetOutputWrapperFieldPath(op)
+```
+
+2. Align `Get` naming and return types in `ackgenconfig` (i.e. Get__ , not Find__)
+  * open to discussion on the exact wording/format as long as it's consistent. Proposing prepending funcs with `Get` and return the desired value and error to start
+
+```
+// Current
+func (c *Config) ResourceConfig(name string) (*ResourceConfig, bool)
+
+// Proposed
+func (c *Config) GetResourceConfig(name string) (*ResourceConfig, error)
+
+```
+
+3. Look for other wrappers and access to `ackgenconfig` and consildate to `config` package
+
+```
+// Current
+func getSortedLateInitFieldsAndConfig(cfg *ackgenconfig.Config, r *model.CRD) ([]string, map[string]*ackgenconfig.LateInitializeConfig)
+
+
+// Proposed
+func (c *Config) GetLateInitFieldsAndConfig(cfg *ackgenconfig.Config, r *model.CRD) (map[string]*ackgenconfig.LateInitializeConfig, error)
+
+// Note, this simply gets the config; sorting and any processing AFTER getting config is done by the client
+
+```
+
+#### Centralize `reconciliation`
+Similar to the above, move `reconcialiation`-related functions from `code` package to `ackmodel` to realign responsibilities.
+
+1. Refactor `type CRD struct` in [crd.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/model/crd.go#L64) to remove `sdkapi`  as attribute and fix breakage
```

Comment:
I can tell you I tried to do this when working on the [`custom_shapes`](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/sdk/custom_shapes.go) work. This is a lot harder than you might originally think. I believe it will take quite a bit of untangling and abstraction to pull off, but it is absolutely worth it to do.